### PR TITLE
Adding compile test and fixing namespace issues

### PR DESCRIPTION
**Pull request by distantcam**
This confirms the generated code compiles without issue in the source generator tests.

Also made the generated code use global namespaces to avoid issues with user types with the same names.

### CODE REVIEW

**Comment by SimonCropp** [comment]
@distantcam thanks

@MattKotsenas can u give this a once over?

File: src/Verify.MSTest.SourceGenerator/Emitter.cs
Comment (MattKotsenas) [inline src/Verify.MSTest.SourceGenerator/Emitter.cs:62]: I'm not sure if I agree with this change. What's the scenario where it's needed?

`new` _hides_ the base member, but it doesn't override it. That means code that accesses the base class as the base type will continue to use the "hidden" accessor. Consider a scenario like this where several test classes share a base class that defines a set of "helpers":

```csharp
public abstract class Base
{
    public TestContext TestContext { get; set; } = null!;

    protected void Helper() => if (TestContext is null) { throw new("TestContext is null"); }
}

[TestClass]
public class Derived : Base
{
    // This line simulates the generator, where it's adding the TestContext property with the `new` keyword
    public new TestContext TestContext { get; set; } = null!;

    [TestMethod]
    public void TestMethod1()
    {
        Helper(); // throws!
        Assert.IsTrue(true);
    }
}
```

This code throws in `Helper()` because MSTest won't do property injection on the base class (correctly because it isn't a `[TestClass]` itself), and Base can't see the TestContext in `Derived` because the base property isn't marked `virtual`.

If there are cases where the generator would be hiding a base member, I'd prefer compilation errors to using `new` and introducing subtle bugs. However, I thought `InheritanceTests` covered all the permutations. Are there scenarios where the generator is hiding the base member?

**Review by MattKotsenas — CHANGES_REQUESTED** [review]
Thanks for this! I appreciate you giving it a try and making improvements!

I wholeheartedly agree with:

- Specifying the full namespace for `TestContext` in the generated code
- Asserting no compilation errors in the driver

Depending on how fast we can iterate I'd be in favor of breaking those parts into their own PR to get them merged quickly (I'll defer to @SimonCropp since I'm not a maintainer and don't have merge capabilities anyways).

I left some comments regarding:

- The reference assembly story
- The use of the `new` keyword to hide base members

take a look there and let me know what you think, or if there's things I'm missing.

Thanks again!

File: src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj
Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj:8]: Yeah this and the package ref were the changes I liked the least. I'm ok with rolling that back as it's not adding any benefits in this case.

File: src/Verify.MSTest.SourceGenerator/Emitter.cs
Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator/Emitter.cs:62]: [This](https://github.com/VerifyTests/Verify/blob/main/src/Verify.MSTest.SourceGenerator.Tests/InheritanceTests.cs#L62-L79) test with an existing `TestContext` in the base class is the test that needed the new.

We just need a way to solve that problem.

File: src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj
Comment (MattKotsenas) [reply src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj:8]: OK then yeah let's do that for now. Eventually we may want / need to run tests per TFM, in which case this code would need to change anyways. If / when that happens is a good opportunity to reconsider the reference assembly code.

File: src/Verify.MSTest.SourceGenerator/Emitter.cs
Comment (MattKotsenas) [reply src/Verify.MSTest.SourceGenerator/Emitter.cs:62]: Ah, well your diagnostics assertions are already paying off then :)

I think this is a case of under specification in the tests on my part. This is the same as the case above, where the user's code is fundamentally ambiguous and needs to be changed.

Thus, I think the previous generated code was correct in the sense that we shouldn't use `new` and causing a CS0108 is the right thing to do in this case.

How does this sound for a plan? We should revert the `HasExistingTestContextProperty` and associated code, and instead update the test so that rather than asserting that there are no diagnostics, assert that for this test CS0108 _should_ exist.

Does that make sense? Let me know if you have any questions!

Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator/Emitter.cs:62]: As mentioned there was 1 test that failed to compile without warnings because of a pre-existing TestContext property, causing this issue.

Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator/Emitter.cs]: Ok, I've left the generator as is, and updated the test to expect the diagnostic error.

File: src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj
Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator.Tests/Verify.MSTest.SourceGenerator.Tests.csproj]: Reverted.

File: src/Verify.MSTest.SourceGenerator.Tests/TestBase.cs
Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator.Tests/TestBase.cs]: Done.

File: src/Verify.MSTest.SourceGenerator/Parser.cs
Comment (MattKotsenas) [inline src/Verify.MSTest.SourceGenerator/Parser.cs:41]: This appears to be dead code now. Can you remove this helper and the corresponding property in `ClassToGenerate`?

If not, what's the use case for it now?

Thanks!

**Comment by MattKotsenas** [comment]
Thanks @distantcam for the updates! Looks like there's just bit of left-over dead code. Mind taking one last look?

@SimonCropp once the update to remove the `HasExistingTestContextProperty` machinery in `Parser` and `ClassToGenerate` is in this looks good to merge from my side. Mind taking a look and letting me know if you agree / have questions / need anything?

Thanks!

File: src/Verify.MSTest.SourceGenerator/Parser.cs
Comment (distantcam) [reply src/Verify.MSTest.SourceGenerator/Parser.cs:41]: Oops yeah. I was experimenting with a possibly different diagnostic, but CS0108 will do.

**Review by MattKotsenas — APPROVED** [review]

**Comment by distantcam** [comment]
All cleaned up and merge conflicts resolved. This is ready to go.

File: src/Verify.MSTest.SourceGenerator.Tests/TestDriver.cs
Comment (SimonCropp) [reply src/Verify.MSTest.SourceGenerator.Tests/TestDriver.cs]: @distantcam thoughts on this?

**Pull request merged**
